File: id_operateur/app/Models/Personne.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class Personne extends Model
{
    use HasFactory;

    protected $fillable = [
        'type', 'nom', 'prenom', 'date_naissance', 'lieu_naissance',
        'nationalite', 'sexe', 'cni_numero', 'denomination_sociale',
        'forme_juridique', 'date_creation', 'adresse', 'quartier',
        'commune', 'ville', 'province', 'telephone', 'telephone_2',
        'email', 'site_web', 'est_actif', 'est_formalise',
        'date_formalisation', 'latitude', 'longitude', 'avatar',
        'id_quartier', 'id_province', 'id_ville'
    ];

    protected $casts = [
        'est_actif' => 'boolean',
        'est_formalise' => 'boolean',
        'date_naissance' => 'date',
        'date_creation' => 'date',
        'date_formalisation' => 'datetime',
        'latitude' => 'decimal:8',
        'longitude' => 'decimal:8',
    ];

    // ============================================================
    // RELATIONS
    // ============================================================
    
    public function utilisateur()
    {
        return $this->hasOne(Utilisateur::class);
    }

    public function quartierRelation()
    {
        return $this->belongsTo(Quartier::class, 'id_quartier');
    }

    public function provinceRelation()
    {
        return $this->belongsTo(Province::class, 'id_province');
    }

    public function villeRelation()
    {
        return $this->belongsTo(Ville::class, 'id_ville');
    }

    public function identifiantsOfficiels()
    {
        return $this->hasMany(IdentifiantOfficiel::class);
    }

    /**
     * Relation avec les activités économiques
     * Spécification explicite des clés pour éviter les erreurs
     */
    public function activites()
    {
        return $this->belongsToMany(
            ActiviteEconomique::class,    // Modèle lié
            'personne_activites',          // Table pivot
            'personne_id',                 // Clé étrangère de la table courante dans la pivot
            'activite_id'                  // Clé étrangère du modèle lié dans la pivot
        )
        ->withPivot('est_principale', 'date_debut', 'date_fin')
        ->withTimestamps();
    }

    public function biensImmobiliers()
    {
        return $this->hasMany(BienImmobilier::class, 'proprietaire_id');
    }

    public function vehicules()
    {
        return $this->hasMany(Vehicule::class, 'proprietaire_id');
    }

    public function permisAutorisations()
    {
        return $this->hasMany(PermisAutorisation::class);
    }

    public function declarationsPaiements()
    {
        return $this->hasMany(DeclarationPaiement::class);
    }

    public function documents()
    {
        return $this->hasMany(Document::class);
    }

    public function notifications()
    {
        return $this->hasMany(Notification::class);
    }

    // ============================================================
    // SCOPES
    // ============================================================
    
    public function scopeActif($query)
    {
        return $query->where('est_actif', true);
    }

    public function scopeFormalise($query)
    {
        return $query->where('est_formalise', true);
    }

    public function scopePhysique($query)
    {
        return $query->where('type', 'physique');
    }

    public function scopeMorale($query)
    {
        return $query->where('type', 'morale');
    }

    // ============================================================
    // MUTATEURS
    // ============================================================
    
    public function getNomCompletAttribute()
    {
        if ($this->type === 'physique') {
            return $this->prenom . ' ' . $this->nom;
        }
        return $this->denomination_sociale;
    }

    public function getAgeAttribute()
    {
        if (!$this->date_naissance) {
            return null;
        }
        return $this->date_naissance->age;
    }

    // ============================================================
    // MÉTHODES UTILITAIRES
    // ============================================================
    
    public function aUnCompteUtilisateur()
    {
        return $this->utilisateur()->exists();
    }

    public function estFormalisable()
    {
        return !$this->est_formalise && $this->est_actif;
    }
}

File: gs_operateur_mobile/lib/models/personne.dart
class Personne {
  final int id;
  final String? type;
  final String nom;
  final String postnom;
  final String prenom;
  final String? denominationSociale;
  final String? formeJuridique;
  final String? sexe;
  final String? telephone;
  final String? email;
  final String? adresse;
  final String? cni;
  final DateTime? dateNaissance;
  final String? lieuNaissance;
  final int? idProvince;
  final int? idVille;
  final int? communeId;
  final int? idQuartier;
  final String? nomComplet;
  final String? avatar;
  final String? statut;
  final DateTime? createdAt;

  Personne({
    required this.id,
    this.type,
    required this.nom,
    required this.postnom,
    required this.prenom,
    this.denominationSociale,
    this.formeJuridique,
    this.sexe,
    this.telephone,
    this.email,
    this.adresse,
    this.cni,
    this.dateNaissance,
    this.lieuNaissance,
    this.idProvince,
    this.idVille,
    this.communeId,
    this.idQuartier,
    this.nomComplet,
    this.avatar,
    this.statut,
    this.createdAt,
  });

  factory Personne.fromJson(Map<String, dynamic> json) {
    return Personne(
      id: json['id'] ?? 0,
      type: json['type'],
      nom: json['nom'] ?? '',
      postnom: json['postnom'] ?? '',
      prenom: json['prenom'] ?? '',
      denominationSociale: json['denomination_sociale'],
      formeJuridique: json['forme_juridique'],
      sexe: json['sexe'],
      telephone: json['telephone'],
      email: json['email'],
      adresse: json['adresse'],
      cni: json['cni'] ?? json['cni_numero'],
      dateNaissance: json['date_naissance'] != null
          ? DateTime.tryParse(json['date_naissance'])
          : null,
      lieuNaissance: json['lieu_naissance'],
      idProvince: json['id_province'],
      idVille: json['id_ville'],
      communeId: json['commune_id'],
      idQuartier: json['id_quartier'],
      nomComplet: json['nom_complet'],
      avatar: json['avatar'],
      statut: json['statut'],
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at'])
          : null,
    );
  }

  String get fullName {
    if (type == 'morale' && denominationSociale != null && denominationSociale!.isNotEmpty) {
      return denominationSociale!;
    }
    return '$nom $postnom $prenom'.trim();
  }

  Map<String, dynamic> toJson() => {
        if (type != null) 'type': type,
        if (nom.isNotEmpty) 'nom': nom,
        if (postnom.isNotEmpty) 'postnom': postnom,
        if (prenom.isNotEmpty) 'prenom': prenom,
        if (denominationSociale != null) 'denomination_sociale': denominationSociale,
        if (formeJuridique != null) 'forme_juridique': formeJuridique,
        if (sexe != null) 'sexe': sexe,
        if (telephone != null && telephone!.isNotEmpty) 'telephone': telephone,
        if (email != null && email!.isNotEmpty) 'email': email,
        if (adresse != null && adresse!.isNotEmpty) 'adresse': adresse,
        if (cni != null && cni!.isNotEmpty) 'cni': cni,
        if (dateNaissance != null)
          'date_naissance': dateNaissance!.toIso8601String().substring(0, 10),
        if (lieuNaissance != null) 'lieu_naissance': lieuNaissance,
        if (idProvince != null) 'id_province': idProvince,
        if (idVille != null) 'id_ville': idVille,
        if (communeId != null) 'commune_id': communeId,
        if (idQuartier != null) 'id_quartier': idQuartier,
      };
}
